/**
 * e8yes demo web server.
 *
 * <p>Copyright (C) 2020 Chifeng Wen {daviesx66@gmail.com}
 *
 * <p>This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * <p>You should have received a copy of the GNU General Public License along with this program. If
 * not, see <http://www.gnu.org/licenses/>.
 */

syntax = "proto3";

package e8;

import "file.proto";

message ChatMessageEntry {
    // The following tuple (thread_id, message_seq_id) constitute a unique ID
    // for each chat message.
    int64 thread_id = 1;
    int64 message_seq_id = 2;

    // ID of the sender of this chat message.
    int64 sender_id = 3;

    // Text content.
    repeated string texts = 4;

    // Read-only media file access tokens.
    repeated FileTokenAccess media_file_accesses = 5;

    // Read-only binary file access tokens.
    repeated FileTokenAccess binary_file_accesses = 6;

    // Creation timestamp.
    int64 created_at = 7;
}

enum ChatMessageThreadType {
    CMTT_INVALID = 0;
    CMTT_TEMPORAL = 1;
    CMTT_POPUP = 2;
}

message ChatMessageThread {
    int64 thread_id = 1;
    string thread_title = 2;
    int64 created_at = 3;
    ChatMessageThreadType thread_type = 4;
    repeated ChatMessageEntry messages = 5;
}
